### PR TITLE
Fix a double free issue when signing SM2 cert

### PR DESCRIPTION
**Pull request by InfoHunter**
If the SM2 ID value has not been passed correctly when signing an SM2 certificate/certificate request, a double free occurs. For instance:

``` 
 openssl req -x509 ... -sm2-id 1234567812345678
```

The '-sm2-id' should not be used in this scenario, while the '-sigopt' is the correct one to use. Documentation has also been updated to make the options more clear.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by InfoHunter** [comment]
Ping @openssl/committers ?

**Comment by slontis** [comment]
I tried looking at the flow for this.. I can see other problems in here also..
e.g:- do_X509_sign()
if EVP_MD_CTX_new returns NULL..
it would fall over in pctx = EVP_MD_CTX_pkey_ctx(mctx);

**Comment by slontis** [comment]
Did you valgrind this to make sure it free correctly?

File: apps/req.c [hunk 1]
```diff
@@ -1734,10 +1734,6 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
 
     ret = 1;
  err:
-#ifndef OPENSSL_NO_SM2
-    if (!ret)
```
Comment (slontis) [inline apps/req.c]: Where is the other free?
If it happens because of EVP_MD_CTX_set_pkey_ctx(ctx, pctx); then you may still need this call fro some paths..

**Comment by InfoHunter** [comment]
> Did you valgrind this to make sure it frees correctly?

Not quite yet, but that could be done

File: apps/req.c (hunk 1 above)
Comment (InfoHunter) [reply apps/req.c]: Any SM2 related `free` should all go in the outer `do_X509_sign`, this should only apply to SM2 keys

Comment (InfoHunter) [reply apps/req.c]: Ahh, I get you point. So it seems the proper way is to handle to outer free instead of this one...

**Comment by InfoHunter** [comment]
Let's me reconsider this part...

File: apps/req.c (hunk 1 above)
Comment (slontis) [reply apps/req.c]: There are multiple issues with this currently.
I dont think you should remove the inner cleanup - If it allocates the data then it should clean up the data if it fails..
The outer one shouldnt just fall thru when there is an error.

Comment (InfoHunter) [reply apps/req.c]: So at the outer cleanup, it should distinguish if the inner call is failed or not, only non-failed inner call should be cleanup at outer space..

Comment (slontis) [reply apps/req.c]: In the inner it should probably be doing...

.....
  EVP_MD_CTX_set_pkey_ctx(ctx, pctx);
  pctx = NULL;

......


if (pctx == NULL)
    EVP_MD_CTX_set_pkey_ctx(ctx, NULL);
else
   EVP_PKEY_CTX_free(pctx);

Comment (InfoHunter) [reply apps/req.c]: It seems this is not necessary. Even if an error occurred after `EVP_MD_CTX_set_pkey_ctx` is called, the original logic can still free the `pctx` and EVP_PKEY_CTX_free accepts NULL input

**Comment by InfoHunter** [comment]
> I tried looking at the flow for this.. I can see other problems in here also..
> e.g:- do_X509_sign()
> if EVP_MD_CTX_new returns NULL..
> it would fall over in pctx = EVP_MD_CTX_pkey_ctx(mctx);

This is also fixed by the latest fixup commit

File: apps/req.c [hunk 2]
```diff
@@ -1744,18 +1744,23 @@ static int do_sign_init(EVP_MD_CTX *ctx, EVP_PKEY *pkey,
...
     rv = do_sign_init(mctx, pkey, md, sigopts);
-    if (rv > 0)
+    if (rv > 0) {
         rv = X509_sign_ctx(x, mctx);
+        try_free = 1;
+    }
 #ifndef OPENSSL_NO_SM2
-    /* only in SM2 case we need to free the pctx explicitly */
-    if (ec_pkey_is_sm2(pkey)) {
+    /*
```
Comment (slontis) [inline apps/req.c]: Probably better not to use try_free variable as it may complain if OPENSSL_NO_SM2 is defined.
Move the code up instead ie. 
if (rv > 0) {
   rv = X509_sign_ctx(x, mctx);
#ifndef OPENSSL_NO_SM2
#endif
}

File: apps/req.c (hunk 1 above)
Comment (slontis) [reply apps/req.c]: The point is if a function allocs memory and then assigns it to something in the MD_CTX, then MD_CTX probably shouldnt be pointing to a deallocated memory location when the pkey ctx is freed. If you use the other function then it frees the pkey and clears the pointer and flags that it uses.

File: apps/req.c (hunk 2 above)
Comment (InfoHunter) [reply apps/req.c]: Sure, no problem

**Comment by InfoHunter** [comment]
New commit pushed to address @slontis 's comment

**Comment by slontis** [comment]
Ahh I see.
EVP_MD_CTX_free() calls EVP_MD_CTX_reset() which does..
    if (!EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX))
        EVP_PKEY_CTX_free(ctx->pctx);

EVP_MD_CTX_set_pkey_ctx() sets the EVP_MD_CTX_FLAG_KEEP_PKEY_CTX..

So ignore my other comment (I didnt consider this wierd code :))

**Comment by InfoHunter** [comment]
> Ahh I see.
> EVP_MD_CTX_free() calls EVP_MD_CTX_reset() which does..
> if (!EVP_MD_CTX_test_flags(ctx, EVP_MD_CTX_FLAG_KEEP_PKEY_CTX))
> EVP_PKEY_CTX_free(ctx->pctx);
> 
> EVP_MD_CTX_set_pkey_ctx() sets the EVP_MD_CTX_FLAG_KEEP_PKEY_CTX..
> 
> So ignore my other comment (I didnt consider this wierd code :))

Right, this is very weird since SM2 requires special option to be set before signing or verification...

**Comment by InfoHunter** [comment]
Furthermore, I think those `do_X509_xxx_sign` functions in apps/req.c should be abstracted to base on some common routine to reduce code duplication. Will do that after this PR is merged.

**Comment by InfoHunter** [comment]
Ping?

File: apps/req.c [hunk 3]
```diff
@@ -1751,15 +1751,19 @@ int do_X509_sign(X509 *x, EVP_PKEY *pkey, const EVP_MD *md,
...
         rv = X509_sign_ctx(x, mctx);
 #ifndef OPENSSL_NO_SM2
-    /* only in SM2 case we need to free the pctx explicitly */
-    if (ec_pkey_is_sm2(pkey)) {
-        pctx = EVP_MD_CTX_pkey_ctx(mctx);
-        EVP_PKEY_CTX_free(pctx);
-    }
+        /*
+         * only in SM2 case we need to free the pctx explicitly
+         * if do_sign_init() fails, pctx is already freed in it
```
[the remark this answers is not in the document]
Comment (InfoHunter) [reply apps/req.c]: Unfortunately the `X509_sign_ctx` requires to use the `pctx` assigned in the `mctx`, so it can only be freed after the `X509_sign_ctx` call.

[the remark this answers is not in the document]
Comment (mattcaswell) [reply apps/req.c]: Hmm. Perhaps then do_sign_init should pass back the pctx (or NULL if it didn't create its own one). Then you can just free it unconditionally

[the remark this answers is not in the document]
Comment (InfoHunter) [reply apps/req.c]: Yes, that was in fact considered when the code was firstly created. But I found that would make the code even harder to understand. Since there is an `EVP_MD_CTX_pkey_ctx` API we could utilize, I think we should use that one.

[the remark this answers is not in the document]
Comment (InfoHunter) [reply apps/req.c]: So...considerations?

[the remark this answers is not in the document]
Comment (mattcaswell) [reply apps/req.c]: Ok - if you have considered it then no problem.

File: doc/man1/openssl-req.pod [hunk 4]
```diff
@@ -342,13 +342,13 @@ for key generation operations.
...
-required by the SM2 signature algorithm for signing and verification.
+Specify the ID string to use when verifying an SM2 certificate request. The ID
+string is required by the SM2 signature algorithm for signing and verification.
 
 =item B<-sm2-hex-id>
 
-Specify a binary ID string to use when signing or verifying using an SM2
-certificate. The argument for this option is string of hexadecimal digits.
+Specify a binary ID string to use when verifying an SM2 certificate request.
+The argument for this option is string of hexadecimal digits.
```
Comment (mattcaswell) [inline doc/man1/openssl-req.pod]: This change seems quite unrelated. Probably should have been a different PR - but at least in a different commit?

Comment (InfoHunter) [reply doc/man1/openssl-req.pod]: No problem, will do that.

Comment (InfoHunter) [reply doc/man1/openssl-req.pod]: Fixed, move the doc change to another commit

**Comment by InfoHunter** [comment]
Looks good now? @mattcaswell

**Comment by InfoHunter** [comment]
Merged to master as dbb7212 and d332997. Thanks for the review, closing...
